# <center>**AI Image Generators Scripts**</center>


# Image Generator: TEST 1

In [16]:
!pip install -q diffusers transformers accelerate safetensors
import torch
from diffusers import StableDiffusionPipeline

model_id = "runwayml/stable-diffusion-v1-5"   # smaller than SDXL
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32)
device = "cuda" if torch.cuda.is_available() else "cpu"
pipe = pipe.to(device)

prompt = "Beautiful skinny chick wearing booty shorts and a pink t-shirt staring out at a dark forest, with the camera behind her (where we can see her sexy, slim body and firm ass from the back)."
image = pipe(prompt, num_inference_steps=20, guidance_scale=7.5).images[0]
image.save("forest.png")
print("Saved forest.png on", device)


Loading pipeline components...:   0%|          | 0/7 [00:00<?, ?it/s]

  0%|          | 0/20 [00:00<?, ?it/s]

Saved forest.png on cpu


---

# Image Generator: TEST 2 (Auto-select model)

In [14]:
# Auto-select model based on GPU availability
!pip install -q diffusers transformers accelerate safetensors
import torch
from diffusers import StableDiffusionPipeline

use_cuda = torch.cuda.is_available()
print("CUDA available:", use_cuda)

if use_cuda:
    # If GPU exists, try a medium model (v1-5). You can switch to SDXL only if you know your GPU has enough VRAM.
    model_id = "runwayml/stable-diffusion-v1-5"
    dtype = torch.float16
    device = "cuda"
    steps = 20
else:
    model_id = "runwayml/stable-diffusion-v1-5"
    dtype = torch.float32
    device = "cpu"
    steps = 15
    print("Warning: Running on CPU will be slow. Consider switching runtime to GPU.")

pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=dtype)
pipe = pipe.to(device)
prompt = "Beautiful girl wearing booty shorts and a pink t-shirt staring out at a dark forest, with the camera behind her (where we can see her sexy, slim body and firm ass from the back). All of her curves are shown while the outline of her cameltoe is seen from her panties."
image = pipe(prompt, num_inference_steps=steps, guidance_scale=7.5).images[0]
image.save("auto_forest.png")
print("Saved auto_forest.png (device:", device, ", steps:", steps, ")")


CUDA available: False


model_index.json:   0%|          | 0.00/541 [00:00<?, ?B/s]

Fetching 15 files:   0%|          | 0/15 [00:00<?, ?it/s]

safety_checker/model.safetensors:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

scheduler_config.json:   0%|          | 0.00/308 [00:00<?, ?B/s]

config.json: 0.00B [00:00, ?B/s]

preprocessor_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/617 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/472 [00:00<?, ?B/s]

text_encoder/model.safetensors:   0%|          | 0.00/492M [00:00<?, ?B/s]

unet/diffusion_pytorch_model.safetensors:   0%|          | 0.00/3.44G [00:00<?, ?B/s]

vae/diffusion_pytorch_model.safetensors:   0%|          | 0.00/335M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/806 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/743 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

config.json:   0%|          | 0.00/547 [00:00<?, ?B/s]

Loading pipeline components...:   0%|          | 0/7 [00:00<?, ?it/s]

  0%|          | 0/15 [00:00<?, ?it/s]

Saved auto_forest.png (device: cpu , steps: 15 )


# Image Generator: TEST 3 (HuggingFace)

### HuggingFace A)

In [9]:
from huggingface_hub import login
login(token="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")

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


In [7]:
# Hugging Face Inference API example (quick)
!pip install -q huggingface_hub pillow
from huggingface_hub import InferenceClient
from PIL import Image
import io
import os

# your full token here
os.environ["HF_TOKEN"] = "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"

client = InferenceClient(
    "stabilityai/stable-diffusion-2-1",
    token=os.environ["HF_TOKEN"]
)

prompt = "Beautiful girl wearing booty shorts and a pink t-shirt staring out at a dark forest, with the camera behind her (where we can see her sexy, slim body and firm ass from the back). All of her curves are shown while the outline of her cameltoe is seen from her panties."

# correct call now:
result = client.text_to_image(prompt)

img = Image.open(io.BytesIO(result))
img.save("beautiful_01.png")
print("Saved beautiful_01.png")
img


StopIteration: 

### HuggingFace B)

In [13]:
# Direct API call using requests — clearer error messages
# New router endpoint version
import requests, os, io
from PIL import Image

HF_TOKEN = os.environ.get("HF_TOKEN")
model_id = "runwayml/stable-diffusion-v1-5"

headers = {"Authorization": f"Bearer {HF_TOKEN}"}

api_url = f"https://router.huggingface.co/hf-inference/models/{model_id}"

payload = {
    "inputs": "Beautiful girl wearing booty shorts and a pink t-shirt staring out at a dark forest, with the camera behind her (where we can see her sexy, slim body and firm ass from the back). All of her curves are shown while the outline of her cameltoe is seen from her panties.",
    "parameters": {
        "guidance_scale": 7.5,
        "num_inference_steps": 25
    }
}

resp = requests.post(api_url, headers=headers, json=payload)

if resp.status_code != 200:
    print("HTTP error", resp.status_code, resp.text[:1000])
else:
    img = Image.open(io.BytesIO(resp.content))
    img.save("forest_router.png")
    print("Saved forest_router.png")
    img


HTTP error 404 Not Found


---

# Image Generator: TEST 4

In [ ]:
!pip install -q diffusers transformers accelerate safetensors

In [4]:
# This will generate in a few seconds on Colab GPU.
from diffusers import DiffusionPipeline
import torch

pipe = DiffusionPipeline.from_pretrained(
    "stabilityai/sdxl-turbo",
    torch_dtype=torch.float16,
    variant="fp16",
)
pipe.to("cuda")

prompt = "Beautiful girl wearing booty shorts and a pink t-shirt staring out at a dark forest, with the camera behind her (where we can see her sexy, slim body and firm ass from the back). All of her curves are shown while the outline of her cameltoe is seen from her panties."
image = pipe(prompt, num_inference_steps=1, guidance_scale=0.0).images[0]

image.save("forest.png")
print("Saved forest.png")

model_index.json:   0%|          | 0.00/685 [00:00<?, ?B/s]

Fetching 18 files:   0%|          | 0/18 [00:00<?, ?it/s]

config.json:   0%|          | 0.00/575 [00:00<?, ?B/s]

scheduler_config.json:   0%|          | 0.00/459 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/586 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/565 [00:00<?, ?B/s]

text_encoder_2/model.fp16.safetensors:   0%|          | 0.00/1.39G [00:00<?, ?B/s]

text_encoder/model.fp16.safetensors:   0%|          | 0.00/246M [00:00<?, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/704 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/855 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/460 [00:00<?, ?B/s]

config.json: 0.00B [00:00, ?B/s]

config.json:   0%|          | 0.00/607 [00:00<?, ?B/s]

vae/diffusion_pytorch_model.fp16.safeten(…):   0%|          | 0.00/167M [00:00<?, ?B/s]

unet/diffusion_pytorch_model.fp16.safete(…):   0%|          | 0.00/5.14G [00:00<?, ?B/s]

Loading pipeline components...:   0%|          | 0/7 [00:00<?, ?it/s]

`torch_dtype` is deprecated! Use `dtype` instead!


RuntimeError: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx

---

# Image Generator: TEST 5

In [1]:
pip install diffusers transformers pillow

In [17]:
!pip install diffusers transformers pillow
from diffusers import StableDiffusionPipeline

# Load the pre-trained model
pipeline = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipeline.to("cpu") # Use "cuda" if GPU is available

# Generate an image
prompt = "Beautiful girl wearing booty shorts and a pink t-shirt staring out at a dark forest, with the camera behind her (where we can see her sexy, slim body and firm ass from the back). All of her curves are shown while the outline of her cameltoe is seen from her panties."
image = pipeline(prompt).images[0]

# Save the image
image.save("output.png")
print("Image saved as output.png")

Loading pipeline components...:   0%|          | 0/7 [00:00<?, ?it/s]

  0%|          | 0/50 [00:00<?, ?it/s]

Image saved as output.png


---

# To Download the Images:

In [ ]:
from google.colab import files
files.download("forest.png")
files.download("output.png")

---

---

# Documents

## ChatGPT response:

Short answer first: yes — your Colab run *did* download the model weights and produced an image; when your notebook printed `Image saved as output.png` it wrote the file into the Colab session filesystem (the notebook's current working directory). In Colab that's almost always `/content/output.png`.

### Where the file is and how to inspect it

Run any of these cells in Colab to confirm and interact with the file:

Check current directory and list files

```python
!pwd
!ls -lah
```

Check that `output.png` exists and show its size

```python
!ls -lh output.png
```

Display the image inline in the notebook

```python
from IPython.display import Image, display
display(Image('output.png'))
```

Download the file to your local PC

```python
from google.colab import files
files.download('output.png')
```

Copy the file to your Google Drive (so it persists across sessions)

```python
from google.colab import drive
drive.mount('/content/drive')     # follow the auth steps
!cp output.png /content/drive/MyDrive/
```

### Why your run took long

* You downloaded very large model files (UNet ~3.4 GB, safety checker, text encoder, VAE, etc.).
* Running stable-diffusion on **CPU** is extremely slow; your log `50/50 [27:37<00:00, 31.33s/it]` shows ~27 minutes for 50 diffusion steps on CPU. That’s expected for a CPU-only Colab runtime.
* If you want much faster runs, switch the Colab runtime to use a GPU (Menu → Runtime → Change runtime type → Hardware accelerator → GPU) and move the pipeline to CUDA with `pipeline.to("cuda")`. Note: GPU availability depends on your Colab plan and session.

### Authentication warning you saw

Colab warned about `HF_TOKEN` because Hugging Face suggests authenticating for model downloads (and some models require a token). To authenticate in the notebook:

```python
from huggingface_hub import login
login(token="hf_xxx...")   # replace with a token from your Hugging Face account
```

Or set it as an environment variable:

```python
import os
os.environ["HF_TOKEN"] = "hf_xxx..."
```

(You can create a token at [https://huggingface.co/settings/tokens](https://huggingface.co/settings/tokens) — do this on Hugging Face’s site.)

### Lighter / faster alternatives & tips

* Use a smaller or optimized model variant (some community models or quantized versions are much smaller).
* Use `torch_dtype=torch.float16` and `pipeline.to("cuda")` to reduce VRAM usage (only on GPU).
* Reduce `num_inference_steps` (e.g., 20–30) for faster results (at slight quality cost).
* Consider using `accelerate`, `safetensors` and `xformers` (for memory-efficient attention) if you move to GPU.

Example setup snippet for GPU (safe, generic; **do not** use with disallowed prompts):

```python
!pip install -q diffusers transformers accelerate safetensors
import torch
from diffusers import StableDiffusionPipeline

pipe = StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",   # example model
    torch_dtype=torch.float16,
    safety_checker=None                # optional; remove if you want the safety checker active
)
pipe = pipe.to("cuda")  # requires GPU runtime
image = pipe("A calm forest at dusk, photorealistic, 35mm, atmospheric").images[0]
image.save("output_safe.png")
```

### Safer prompt guidance

If your goal is to learn prompt engineering, try prompts that describe adult, non-sexualized scenes or objects. Example:

* `A photorealistic portrait of an adult hiker looking at a misty forest at sunrise, cinematic lighting, 35mm lens`
* `A fantasy landscape, towering bioluminescent trees and a small cabin, ultra-detailed`

### If you want, next steps I can help with (safe):

* Show how to switch Colab to GPU and update your notebook to use CUDA.
* Provide a concise, tested Colab notebook cell order that installs dependencies, logs into Hugging Face, loads an optimized pipeline, and saves/downloads images.
* Suggest smaller or quantized models you can use on limited hardware (and provide code to run them).

Tell me which of the above you want next (e.g., “make a GPU-ready Colab cell sequence” or “give me a safe example notebook I can paste into Colab”) and I’ll provide the full cells you can copy-paste.


---

## ChatGPT Response 2:

Thanks — that output confirms there’s no GPU attached to your Colab session right now (`nvidia-smi not found` and `CUDA available: False`). No worries — I’ll give you a clear, copy-paste set of steps and notebook cells you can run *right now* to:

* switch the runtime to GPU correctly,
* re-check GPU availability,
* fall back to a fast, smaller model if a GPU still isn’t available,
* and an alternative path that uses the Hugging Face Inference API (fast + no GPU needed).

Run the items in order. I include short explanations so you know what each cell does.

---

## 1 — Make sure your notebook is set to GPU

In Colab menu:
`Runtime → Change runtime type → Hardware accelerator → GPU → Save`

Then: `Runtime → Restart runtime`

(If you already did that and it still shows no GPU, Colab didn’t assign you a GPU this session — try again later or consider upgrading to Colab Pro.)

---

## 2 — Re-run the GPU check cell (copy/paste into a cell and run)

```python
# Cell A — GPU / PyTorch quick check
!nvidia-smi || echo "nvidia-smi not found (no NVIDIA GPU visible)"
import torch
print("torch version:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())
print("CUDA device count:", torch.cuda.device_count())
if torch.cuda.is_available():
    print("Device name:", torch.cuda.get_device_name(0))
```

If this now prints `nvidia-smi` info and `CUDA available: True`, great — proceed to the GPU pipeline cell below. If it still prints `not found` / `False`, continue to the fallback cells.

---

## 3 — If GPU is available: fast pipeline (copy/paste the two cells)

```python
# Cell B1 — install lightweight dependencies (GPU path)
!pip install -q diffusers transformers accelerate safetensors
```

```python
# Cell B2 — load smaller, reliable SD model and generate (GPU)
import torch
from diffusers import StableDiffusionPipeline
device = "cuda" if torch.cuda.is_available() else "cpu"
print("Using device:", device)

# Smaller model that usually fits common Colab GPUs
model_id = "runwayml/stable-diffusion-v1-5"

# load with half precision on GPU for memory savings
dtype = torch.float16 if device == "cuda" else torch.float32
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=dtype)
pipe = pipe.to(device)

prompt = "Dark forest, cinematic, foggy, dramatic lighting, ultra-detailed"
image = pipe(prompt, num_inference_steps=20, guidance_scale=7.5).images[0]
image.save("forest_gpu.png")
print("Saved forest_gpu.png")
from IPython.display import display, Image
display(Image("forest_gpu.png"))
```

Notes: use `num_inference_steps=15-25` for a good speed/quality balance. If you run out-of-memory on GPU, reduce steps or use a lower-resolution image (the pipeline defaults to 512×512).

---

## 4 — If GPU is NOT available: CPU fallback (smaller model + fewer steps)

This will work without GPU but is slower. Use the smaller `v1-5` model and fewer steps.

```python
# Cell C — CPU fallback (slow but works)
!pip install -q diffusers transformers safetensors
import torch
from diffusers import StableDiffusionPipeline
device = "cpu"
print("Running on CPU (this will be slow).")

pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5")
pipe = pipe.to(device)

prompt = "Dark forest, cinematic, fog, dramatic lighting"
image = pipe(prompt, num_inference_steps=15, guidance_scale=7.0).images[0]
image.save("forest_cpu.png")
print("Saved forest_cpu.png")
from IPython.display import display, Image
display(Image("forest_cpu.png"))
```

Expect several minutes per image on CPU (depends on Colab VM power). Reducing `num_inference_steps` and image size speeds it up but reduces quality.

---

## 5 — Fast alternative: Use the Hugging Face Inference API (no GPU needed, fast)

This uses Hugging Face hosted inference — quick and easy. You’ll need an `HF_TOKEN` (free tier available). Steps:

1. Get a token at [https://huggingface.co/settings/tokens](https://huggingface.co/settings/tokens) (create an account if needed).
2. In the notebook, set it (don’t expose tokens in shared notebooks):

```python
# Cell D1 — Hugging Face Inference API example (quick)
!pip install -q huggingface_hub pillow
from huggingface_hub import hf_hub_download, HfApi, InferenceClient
from PIL import Image
import io
import base64
import os

# Set your token here (or better: use Colab secrets)
os.environ["HF_TOKEN"] = "hf_xxx_REPLACE_WITH_YOUR_TOKEN"

client = InferenceClient(token=os.environ["HF_TOKEN"])

prompt = "Dark forest, cinematic, atmospheric fog, photorealistic"
# model below is an example hosted model — you can replace with other hosted endpoints
result = client.text_to_image(model="stabilityai/stable-diffusion-2-1", input=prompt, wait_for_model=True)

# result is raw bytes; save to file
img = Image.open(io.BytesIO(result))
img.save("forest_hfapi.png")
print("Saved forest_hfapi.png")
display(img)
```

Pros: quick, no GPU setup. Cons: usage rates / quotas may apply; large-scale usage can cost money.

---

## 6 — Tips to get faster & more reliable results in Colab

* Switch runtime to GPU and restart runtime (single most important step).
* Prefer smaller models like `runwayml/stable-diffusion-v1-5` on free GPUs (less VRAM).
* Use `torch.float16` on GPU to reduce memory.
* Use `num_inference_steps=15-25` instead of 50 for faster runs.
* Use `pipe.scheduler = ...` — advanced; leave for later.
* If you need VERY fast low-latency images, use the Hugging Face Inference API or paid services (Replicate, Stability.ai hosted API).

---

## 7 — Extra: automatic notebook that handles both cases

If you want one cell that attempts to detect GPU and pick the best route, use this block (copy/paste entire cell):

```python
# Cell E — Auto pick GPU/CPU and generate with best defaults
!pip install -q diffusers transformers accelerate safetensors huggingface_hub pillow
import torch, os
from diffusers import StableDiffusionPipeline
from IPython.display import display, Image

use_cuda = torch.cuda.is_available()
print("CUDA available:", use_cuda)

model_id = "runwayml/stable-diffusion-v1-5"
dtype = torch.float16 if use_cuda else torch.float32
device = "cuda" if use_cuda else "cpu"
steps = 20 if use_cuda else 15

print(f"Loading {model_id} as dtype={dtype} on {device} (steps={steps}) — this may take a bit.")
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=dtype)
pipe = pipe.to(device)

prompt = "Dark forest, moody fog, dramatic lighting, cinematic"
image = pipe(prompt, num_inference_steps=steps, guidance_scale=7.5).images[0]
outname = "forest_auto_gpu.png" if use_cuda else "forest_auto_cpu.png"
image.save(outname)
print("Saved", outname)
display(Image(outname))
```

---

## 8 — If Colab still won’t give you a GPU

* Try a new notebook or open an incognito browser window and start a fresh Colab session.
* `Runtime → Manage sessions` → terminate other sessions to free quota.
* Keep trying or consider Colab Pro for higher GPU availability.
* Or use Hosted APIs (Hugging Face or Replicate) for immediate speed.

---

## 9 — If you’d like, I can:

* Paste a full 6–8 cell Colab notebook (ready to paste) that: installs deps, checks GPU, generates, displays, and offers a download button — done and tested for beginner-friendly flow.
* Show how to mount your Google Drive and automatically copy every generated image there.
* Show how to use SDXL Turbo or other "faster" models *if* you get a GPU with enough VRAM.

Which of those would you like next? I recommend first: try the `Cell A` check after switching runtime to GPU — paste the output here if it still shows no GPU and I’ll walk you through the next troubleshooting steps.
